### PR TITLE
testlib improvements

### PR DESCRIPTION
**Pull request by zmiklank**
1. kill image build after ten minutes

    The image build is run in subshell, which is killed if does not finish
    after 10 minutes. If the build is finished sooner no action is taken.

    This could however cause problems when sending SIGINT to the test
    process, as trap is not inherited by subshells.

2. use different trap functions for EXIT and SIGINT for container tests

    so when the tests are stopped with SIGINT the final tasks are minimized
    and the waiting time for results is reduced. That means, that image size
    and stats are not shown in this case.

3. OCP tests: examine the exit code in the trap function

    in OCP3 tests we still use set -e for discovering test failure. However,
    if the non-zero exit code is caught by trap function the whole bash ends
    with exit code from the trap function itself and not with the exit code
    with the tests.

    This commit solves the issue by checking what value was returned from
    the testsuite and exits the whole testsuite based on that.

### CODE REVIEW

**Comment by zmiklank** [comment]
[test]

**Comment by zmiklank** [comment]
[test-all]

**Comment by zmiklank** [comment]
[test-all]

**Review by pkubatrh — CHANGES_REQUESTED** [review]
Added some comments.

Additionally, one of the commits says "ct_cleanup now does not return TESTUITE_RESULT, it only sets it." but from what I can see, `ct_cleanup` does not set `TESTUITE_RESULT` in any way, so the message could be confusing to the reader.

File: test-lib.sh [hunk 1]
```diff
@@ -1327,7 +1364,6 @@ ct_run_tests_from_testset() {
     fi
     time_diff=$(ct_timestamp_diff "$time_beg" "$time_end")
     printf -v TEST_SUMMARY "%s %s for '%s' %s (%s)\n" "${TEST_SUMMARY:-}" "${test_msg}" "${app_name}" "$test_case" "$time_diff"
-    [ -n "${FAIL_QUICKLY:-}" ] && return 1
```
[the remark this answers is not in the document]
Comment (zmiklank) [reply test-lib.sh:1320]: It is related. Explanation:

This PR introduces checking of return value with which the trap on EXIT function is called, so that we could see if some weird (not expected) failures happen in the tests.
But if [ -n "${FAIL_QUICKLY}" ] is evaluated as not true (which is always, because we do not use it at all) and is also the last command of the shell (which is in except of ruby container tests always) then the shell ends with 1 and testsuite ends with FAILURE, as the return value is caught and evaluated by the trap function.

I counted with a fact, that this functionality is not used and that we would need to introduce something more robust during the `prepare` functions that we have in few of our containers' tests anyway. So I removed that.
However, if you think that this line should be kept in the code I will rework the solution.

File: test-lib.sh [hunk 2]
```diff
@@ -195,7 +213,30 @@ function ct_show_results() {
...
+  trap ct_trap_on_sigint SIGINT
+}
+
+# ct_trap_on_exit
+# --------------------
+function ct_trap_on_exit() {
+  local exit_code=$?
+  [ $exit_code -eq 130 ] && return # we do not want to catch SIGINT here
+  echo "Tests finished with EXIT=$exit_code"
+  exit_code=$(("${TESTSUITE_RESULT:-0}"+exit_code))
```
[the remark this answers is not in the document]
Comment (zmiklank) [reply test-lib.sh:226]: We should not really care about what the script returns, as now all the relevant `exit` calls are removed from the tests.

I put it there for future, just to be sure that we catch some not expected behavior, if any is added.

There should not be any additional problems with this line as I see it. But sure, I can add some explanatory comment. Or remove the functionality at all, if you think it is unnecessary.

[the remark this answers is not in the document]
Comment (pkubatrh) [reply test-lib.sh:226]: I think we can keep it as is but a comment would be nice. The only thing I would be afraid of is the availability of this construct in some earlier versions of shells, since we have seen some operating systems (especially Mac) to be stuck in the past. :) But we can come back to this if we get complaints.

**Comment by zmiklank** [comment]
[test-all]

File: test-lib.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (phracek) [reply test-lib.sh:1320]: If we have a variable like `FAIL_QUICKLY` defined, but never used, I am fine with removal.

File: test-lib.sh [hunk 3]
```diff
@@ -60,39 +62,12 @@ function ct_init() {
...
-    docker rm -v "$container"
-    rm -f "$cid_file"
-  done
-
-  rm -rf "$CID_FILE_DIR"
-  exit "${TESTSUITE_RESULT:-0}"
+  ct_clean_containers
 }
 
 # ct_build_image_and_parse_id
```
Comment (phracek) [inline test-lib.sh:73]: Personally, this function becomes complicated from my point of view. As I am not a bash expert when the build took a lot of time? It would be nice to have a description or explanation of what the code does. Even for basic users.

**Review by phracek — CHANGES_REQUESTED** [review]
Please address my comments.

File: test-lib.sh (hunk 3 above)
Comment (zmiklank) [reply test-lib.sh:73]: The problem appeared with ruby. One of the build steps got cycled and the build never ended, the CI timeouted.
There weren't any logs too, as they are printed just after the image is build.

But sure, I can add some explanatory comments.

**Comment by zmiklank** [comment]
[test]

File: test-lib.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (zmiklank) [reply test-lib.sh:239]: Actually, this problem already stroke on CentOS7. So I need to rework it. Thanks for catching it :).

**Comment by zmiklank** [comment]
[test]

File: test-lib.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (zmiklank) [reply test-lib.sh]: Fixed with: https://github.com/sclorg/container-common-scripts/pull/338/commits/c0499b7b040b23fc004cfd90bd2f02be7d079d82.

**Review by phracek — CHANGES_REQUESTED** [review]
in build.sh we have a similar function, why do not use the same approach? See https://github.com/sclorg/container-common-scripts/blob/master/build.sh#L51 and https://github.com/sclorg/container-common-scripts/blob/master/build.sh#L24

File: test-lib.sh [hunk 4]
```diff
@@ -109,15 +84,34 @@ ct_build_image_and_parse_id() {
...
+  local pid_sleep
+  local sleep_time
   log_file="$(mktemp)"
-  ret_val=1
+  sleep_time="10m"
   [ -n "$1" ] && dockerfile="-f $1"
   command="$(echo "docker build --no-cache $dockerfile $2" | tr -d "'")"
-
-  if $command > "$log_file" 2>&1; then
+  # running command in subshell, the subshell in background, storing pid to variable
```
[the remark this answers is not in the document]
Comment (zmiklank) [reply test-lib.sh:94]: Yes, I already explained it here:
https://github.com/sclorg/container-common-scripts/pull/338#discussion_r1142117196

I can explain further if this explanation is not sufficient.

[the remark this answers is not in the document]
Comment (phracek) [reply test-lib.sh:94]: In case of some container build, like `s2i-ruby-container` takes a long time, then we should fix the corresponding container and set up a hard timeout. What about if time out exceeds more than 10m e.g. in the case of s2i-python-container, we will increase the timeout again? The rest of this pull request is fine, but this part of code brings me a lot of issues in the future. What about to add also into test suite s2i-ruby-container? I would not complicate container build so much. Let's build it.

**Review by phracek — CHANGES_REQUESTED** [review]

**Comment by phracek** [comment]
I totally missing tests in this repo that could test corner cases like it is introduced by this pull request.

File: test-lib.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (zmiklank) [reply test-lib.sh:94]: https://github.com/sclorg/s2i-ruby-container/issues/469 This is the issue causing never ending build.

[the remark this answers is not in the document]
Comment (pkubatrh) [reply test-lib.sh:94]: We should really have some timeout, even though 10m is rather arbitrary and I expect we will have to increase it at some point after we hit longer builds for some reason.
Without the timeout we are just hogging testing farm resources without good reason if we hit such an issue at buildtime.

[the remark this answers is not in the document]
Comment (zmiklank) [reply test-lib.sh:94]: As this is the last opened issue I see on this PR, can I now merge it?

[the remark this answers is not in the document]
Comment (phracek) [reply test-lib.sh:94]: Let's get merge it.

**Review by phracek — APPROVED** [review]
